### ISSUE TITLE
str.KLEE memory model

### ISSUE THREAD

**Issue by kren1** [OrenGitHub/klee#18]
I've been thinking  about how to represent abstract buffers in KLEE in relation to how arrays of bitvectors are currently presented.  First some background as to how I understand klee deals with bitvectors and how we do abstract buffers at the moment. I'll consider the symbolic cases, because if all offsets are known  `ObjectState` can directly do read/writes without the solver, so we don't need to think about it.

I'll use the following example:
```C
int i,j = //symbolic
char a[2] = {'h', 'j'}; // represented as klee::Array named a
a[i] = 'g';
a[j] = 0;
char c = a[1];
```

### BV arrays
Bitvectors arrays are represented with the `UpdateList` Object, which is a list of all writes into an `Array` objects. So the `UpdateList` for the example would be something like:

```   ([UpdateNode(i, 'g'), UpdateNode(j, 0)], a) ```

Which gets translated into something like this for the solver expression:

 ```  Write(i, 'g', Write(j, 0, a)) ```

The `Array` (`a` in the example) can either be a concrete array or a symbolic Array. The `a` in the example is a concrete array, with values 'h' and 'j'. If there was a call to `klee_make_symbolic(a, sizeof a)`, it would become concrete.

This way reads simply become a ReadExpr at index of update list:
```   ReadExpr(1, ([UpdateNode(i, 'g'), UpdateNode(j, 0)], a))```
 
or 
 ``` Read(1, Write(i, 'g', Write(j, 0, a))) ```

### Current Abstract Buffers

Currently in strKLEE we  have a StrVarExpr for , `char a[2]`  representing an abstract buffer with the name *AB_a_0*  and length 2. Ignoring the initial constant values, because we don't deal with them at the moment, the way we do writes is to add StrLen and SubString constraints to the **state**. So the two writes would result in something like this constraints:

```
Strlen(AB_a_0) == Strlen(AB_a_1);
AB_a_1 == Substring(0,i, AB_a_0) ++ 'g' ++  Substring(i+1, Strlen(AB_a_0) , AB_a_0);
AB_a_2 == Substring(0,j, AB_a_1) ++ '\0' ++  Substring(j+1, Strlen(AB_a_0) , AB_a_1);
```

The read then simply becomes a StrCharAt expression of the most recent StringVarExpr for the memory object, in this case:

```
StrCharAt(1, AB_a_2)
```

I think there are several issues with this, mostly because it doesn't fit nicely in KLEE and therefore we have to re-implement a lot of existing things:
* Can't use existing functionality to get concrete values for strings
* Doesn't take advantage of `ObjectState` essentially "caching" reads and writes with concrete offsets
* Can't think of anything else now, but I feel it's cumbersome to work with.

### Proposal: Represent Strings as `Arrays`

Because we give AbstractBuffers a size I see no difference between a String and an Array of bitvectors, so I think AbstractBuffers should be represented with the Array class in KLEE, with a special type. I'm thinking Int -> String.  Then basically there should be no change from baseline KLEE in how reads/writes are represented, we just need to make sure that it type checks. The read and writes would be represented exactly the same in klee as they are for BV arrays:

 ``` ReadExpr(1, ([UpdateNode(i, 'g'), UpdateNode(j, 0)], AB_a)) ```

However, they would have different types, so 1,i,j would be an Int type and 'g' and '\0' would be String types. The types would obviously propagate up, so ReadExpr would have String type.When we would translate those into Z3 queries, they would become something like:

``` 
StrCharAt(1, 
     write1 = Substring(0,i, AB_a) ++ 'g' ++  Substring(i+1, Strlen(AB_a) , AB_a)
               in Substring(0,j, write1) ++ '\0' ++ Substring(j+1, Strlen(AB_a, write1)
                )
```

This would replicate the read/write mechanics of BV arrays to strings. It's probably just slower then keeping everything in bitvectors. Where we have the edge is in the string functions/expressions, which would take an entire array as input. This is something KLEE doesn't do for bitvectors, but I think it would work well with strings by having the big substring subexpressions instead of StringVarExprs. That could be slower, which is the only potential downside I see of this approach.

On the other hand the advantages are:
* Dealing with constants is trivial, just replace the AB_a with a string constants
* Get the folding of constant writes for free (courtesy of `ObjectState`)
* Decrease complexity of the code: not need for StrVarExpr and StrConstExpr
* Which means that getting values is easy to integrate in current klee design, because we just associate a vector of unsigned chars to the Array, no need to reinvent the wheel for strings.
* Very easy switching between String world and Bitvector world (comes down to making things type check)
* Probably easy optimisations of Read->ConvertBV->Write->ConvertString->Read pattern
* Hopefully avoiding yet unseen problems due to different abstraction for strings

@OrenGitHub @ccadar  What do you think? Have I missed/misunderstood something important? Is it worth doing? I actually think it isn't that hard, now that I've written this up.

**Comment by OrenGitHub** [comment]
Will look at it this evening thanks -- over the last 4 days I've looked at tens of relevant CVE's so I'd really like to share with you some observations about that too

**Comment by OrenGitHub** [comment]
First of all, I agree that (ideally) str.KLEE should aspire to be a native part of KLEE,
be engineered properly, and follow vanilla.KLEE's established advantages.
I also like the idea of cashing. If we can find examples of consecutive writes and make them a single sub-string constraint - even better. I'm not sure how this representation will enable BitVec <----> String World, I mean if there **aren't any forks** then I agree -- but otherwise, conversion is still complicated. 

Having said that, I think that we should be very CVE oriented.
From my point of view, changing the code should only come (at this stage)
from a bottleneck in a concrete CVE that will be made feasible with that change
Otherwise we won't be able to make the April 16th deadline ...

I'll post my conclusions about CVE's tried so far in a separate issue ...

**Comment by kren1** [comment]
Is the current versions good enough to do a nice evaluation? From what I've seen in other issues, we still seem to be struggling. I don't think tunnel visioning on CVEs is a good idea, when there are known issues, generally I don't advocate for taking engineering shortcuts in the name of "saving time".

**Comment by OrenGitHub** [comment]
makes sense ... but this change will effect every single transformer right? I also don't understand why StrVarExpr / StrConstExpr will not be needed? Basically I'm just a bit afraid this will cause big changes ...

**Comment by kren1** [comment]
Well it effects the start of every transformer, where we replace the StrVarExpr, but I don't see any fundamental changes.

The reason we wouldn't need StrVarExpr any more is that it simply becomes, and `UpdateList([], a)` where a is a symbolic array of type string. Similarly StrConstExpr becomes `UpdateList([], a)` where a is a constant array of String type with values corresponding to the constant string.

Of course it causes reasonably large changes, but I think we are still at the stage where large changes are needed. I don't really see a reason to lock down the code now, unless we can already do a nice evaluation.

**Comment by kren1** [comment]
I just want to document another potential problem with the current approach.

The AB name information is stored in MemoryObjects, which are across states. I think this could cause potential issues, if AB is incremented in a state but not in another one.

**Comment by OrenGitHub** [comment]
Wait but the lastVerion has to be kept in the execution state ... this is what was originally ...
